### PR TITLE
create test script for lion optimizer

### PR DESCRIPTION
**Pull request by DiCeyIII**

### CODE REVIEW

**Comment by DiCeyIII** [comment]
Parallel Training of LLMs Mid Term Project for Okechukwu Okeke and Joshua Udobang

File: include/nntile/kernel/lion_step/cuda.hh [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#include <nntile/base_types.hh>
+#include <cuda_runtime.h>
+
+namespace nntile::kernel::adam_step
+{
+
+template<typename T>
+void cuda(cudaStream_t stream, Index num_iter, Index num_elems, Scalar beta_1,
+        Scalar beta_2, Scalar eps, Scalar lr, Scalar weight_decay,
+        const T *grad, T *first_moment, T *second_moment, T *p)
```
Comment (Muxas) [inline include/nntile/kernel/lion_step/cuda.hh:26]: remove second moments. Rename adam into lion

File: include/nntile/starpu.hh [hunk 2]
```diff
@@ -22,9 +22,8 @@
 #include <nntile/starpu/accumulate_hypot.hh>
 #include <nntile/starpu/accumulate_maxsumexp.hh>
 #include <nntile/starpu/axpy.hh>
-#include <nntile/starpu/add_slice_inplace.hh>
 #include <nntile/starpu/add_slice.hh>
-#include <nntile/starpu/add_fiber_inplace.hh>
```
Comment (Muxas) [inline include/nntile/starpu.hh:27]: And again many headers are deleted. Seems like you started with some old version of NNTile.

File: include/nntile/tensor.hh [hunk 3]
```diff
@@ -21,9 +21,8 @@
 
 // Tensor operations
 #include <nntile/tensor/axpy.hh>
-#include <nntile/tensor/add_slice_inplace.hh>
 #include <nntile/tensor/add_slice.hh>
-#include <nntile/tensor/add_fiber_inplace.hh>
```
Comment (Muxas) [inline include/nntile/tensor.hh:26]: this header also deletes some includes

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -88,11 +87,9 @@ else()
             ${KERNEL_SRC}
             "kernel/accumulate_maxsumexp/cuda.cu"
             "kernel/add/cuda.cu"
-            "kernel/add_slice_inplace/cuda.cu"
```
Comment (Muxas) [inline src/CMakeLists.txt:91]: some sources are deleted. I am sure you did something very strange with git. You forked the repository and then uploaded old version of some files on top of cloned repo. This is not how you work with git.

File: src/kernel/lion_step/cuda.cu [hunk 5]
```diff
@@ -0,0 +1,115 @@
+/*! @copyright (c) 2022-present Skolkovo Institute of Science and Technology
+ *                              (Skoltech), Russia. All rights reserved.
+ *                 2023-present Artificial Intelligence Research Institute
+ *                              (AIRI), Russia. All rights reserved.
+ *
+ * NNTile is software framework for fast training of big neural networks on
+ * distributed-memory heterogeneous systems based on StarPU runtime system.
+ *
+ * @file src/kernel/adam_step/cuda.cu
```
Comment (Muxas) [inline src/kernel/lion_step/cuda.cu:9]: this entire file is just an unchanged adam step, why would one put it onto pull request?

File: src/kernel/lion_step/cpu.cc [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    for (Index i = 0; i < num_elems; ++i)
+    {
+        // Read current parameter & gradient
+        Y p_val    = static_cast<Y>(p[i]);
+        Y grad_val = static_cast<Y>(grad[i]);
+
+        // Apply weight decay if needed
+        if (wd != Y(0))
+        {
+            grad_val += wd * p_val;
```
Comment (Muxas) [inline src/kernel/lion_step/cpu.cc:69]: weight decay is applied incorrectly here

File: src/kernel/lion_step/cpu.cc [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+ */
+{
+    using Y = typename T::repr_t;  // e.g. float or double at runtime
+
+    // Convert scalars to underlying type for faster arithmetic
+    const Y beta_1      = static_cast<Y>(beta_1_);
+    const Y beta_2      = static_cast<Y>(beta_2_);
+    const Y lambda_val  = static_cast<Y>(lambda_);
+    const Y lr          = static_cast<Y>(lr_);
+    const Y wd          = static_cast<Y>(weight_decay_);
```
Comment (Muxas) [inline src/kernel/lion_step/cpu.cc:58]: Parameter lambda is the weight decay, no need to pass it twice. I see, that you did not understand logic of weight decay correctly, that is why you have 2 of them.

File: test_lion_step.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        beta2 (float): Coefficient for updating the momentum m.
+        weight_decay (float): L2 regularization coefficient.
+        num_iter (int): Current iteration number (not used in Lion update, maintained for API symmetry).
+        
+    Returns:
+        tuple: Updated parameters p and momentum m.
+    """
+    # Optionally apply weight decay (L2 regularization) by modifying the gradient.
+    if weight_decay != 0:
+        grad = grad + weight_decay * p
```
Comment (Muxas) [inline test_lion_step.py:43]: this weight decay is incorrect

File: wrappers/python/nntile/functions.py [hunk 9]
```diff
@@ -76,14 +76,6 @@ def gemm_async(
         core_tensor.gemm_async_fp32_fast_tf32(
             alpha, trans_A, A, trans_B, B, beta, C, ndim, batch_ndim, redux
         )
-    elif type(A) is core_tensor.Tensor_fp32_fast_fp16:
```
Comment (Muxas) [inline wrappers/python/nntile/functions.py:79]: Lots of deleted lines in this file mean only one thing -- you messed up with versions.

File: wrappers/python/nntile/nntile_core.cc
Comment (Muxas) [inline wrappers/python/nntile/nntile_core.cc:1]: Lots of deleted lines mean that you have failed with git repository revisions.
